src/java/org/jivesoftware/openfire/pubsub/PubSubEngine.java

```diff
@@ -543,7 +543,7 @@ private void subscribeNode(PubSubService service, IQ iq, Element childElement, E
             return;
         }
         // Check if the subscriber is an anonymous user
-        if (!UserManager.getInstance().isRegisteredUser(subscriberJID)) {
+        if (!UserManager.getInstance().isRegisteredUser(subscriberJID) && !isComponent(subscriberJID)) {
```

Comment:
Assuming Java does shortcuts, you'll want this as the first condition:

if (!isComponent(subscriberJID) && ...)

isRegisteredUser() will do a disco#info query (rather surprisingly) in order to find out whether a remote user is anonymous or not. Not sure this is right, but it's what's done.